azure/services/aso/aso.go

```diff
@@ -202,8 +204,15 @@ func (r *reconciler[T]) CreateOrUpdateResource(ctx context.Context, spec azure.A
 	}
 	parameters.SetAnnotations(annotations)
 
-	diff := cmp.Diff(existing, parameters)
-	if diff == "" {
+	log.V(2).Info("creating or updating resource", "diff", cmp.Diff(existing, parameters))
+	result, err := controllerutil.CreateOrPatch(ctx, r.Client, resource, func() error {
```

Comment:
Now I'm wondering based on @matthchr's comment here https://github.com/kubernetes-sigs/cluster-api-provider-azure/pull/4304#discussion_r1462492672 whether this helper is appropriate to use here or not.

The way things are currently, we capture our intent to either create or update once in the first GET (on line 90), but here the `CreateOrPatch` helper does an additional GET to determine whether to create or update. That means if a user creates a BYO ASO resource in between these two GETs, CAPZ could update that resource when it shouldn't. Today, we would instead see that the resource doesn't exist in the first GET, try to create, and then presumably hit an error saying that the resource already exists, requeue, then eventually find that the resource already exists.

I might see if there's another low-hanging fruit way to shunt some of this create or update logic into its own function without doing another GET.